### ISSUE TITLE
modules should be able to delay initial RUNNING state

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#2010]
A service module may have a complex startup procedure where it waits for several kvs lookups or requests to other services before the module should be considered "initialized".

For instance a scheduler may need to gather configuration, lookup some values from the KVS, set up monitoring of resources, register with job-manager, etc, and only then should it be considered in the active or RUNNING state.

I think currently the module is marked as ready/running once the main handle reactor is entered. This means that the module has to either use a separate reactor (and handle?) or do synchronous calls to perform these initialization routines (unless it doesn't matter if the module is marked running before initialization is complete). 

In either case, it would be less awkward if the module had a way to disable the automatic ready flag, and instead could trigger or enable it manually with an API call.

**Comment by garlick** [comment]
`flux module load` doesn't return (`cmb.insmod` response delayed) until the module leaves INIT state, so the ability to delay leaving INIT can be used to delay starting the initial program before all services are ready, or to prevent a dependent service from loading until its dependency is ready in `rc1`.

What about something like
```c
void flux_module_initialize_incr (flux_t *h);
void flux_module_initialize_decr (flux_t *h);
```

as a sort of semaphore on initialization?  If you don't call it, then state leaves INIT once reactor is entered.  If you do call it, then state leaves INIT once the number of `decr` and `incr` calls balance out?

State is currently reported to the broker using keepalive messages generated by prep/check watchers in `src/broker/modservice.c`.  This would just add a bit of logic so that those are no-ops as long as this count > 0.   The count could be a part of `modservice_ctx_t` in that source file.

**Comment by grondo** [comment]
This would be perfect! Because of the use case, would `flux_module_init_hold (h)` and `flux_module_init_release (h)` work better? Otherwise I'd keep with our other reference naming and suggest `flux_module_init_incref()` `flux_module_init_decref()`

**Comment by garlick** [comment]
That was my first thought, but then I wondered:  if initialization had some optional components that complete asynchronously (say different on a leader vs follower), it might be nice to have a counter built in rather that have to manage one?  It's no more difficult to implement than a "latch flag".

Your names are better!

**Comment by grondo** [comment]
> That was my first thought, but then I wondered: if initialization had some optional components that complete asynchronously (say different on a leader vs follower), it might be nice to have a counter built in rather that have to manage one?

Yes, good point the counter should definitely be built in. I was thinking multiple holds could be placed, each requiring a release. However, if there is some semantics built in to "hold" "release" I'm fine with the incref decref (it just isn't as clear what it does from the function name IMO, but the difference is negligible)

**Comment by garlick** [comment]
Oh, sorry, sounds like we're on the same page.  I'm fine with hold and release with a little documentation on how they are to be used.   Those names do seem clearer now that you mention it.
